TLM/TLM/UI/SubTools/LaneConnectorTool.cs

```diff
@@ -283,84 +275,206 @@ private class LaneEnd {
-                if (!viewOnly && (GetSelectionMode() == SelectionMode.None)) {
-                    Highlight.DrawNodeCircle(
-                        cameraInfo: cameraInfo,
-                        nodeId: (ushort)nodeId,
-                        color: isNodeVisible ? DefaultLaneEndColor : DefaultDisabledLaneEndColor,
-                        alpha: true);
-                }
+            if (!viewOnly) {
+                RenderNodeCircles(cameraInfo);
+            }
 
-                if (!hasMarkers || !isNodeVisible) {
-                    continue;
-                }
+            RenderLaneCurves(cameraInfo);
 
-                float intersectionY = Singleton<TerrainManager>.instance.SampleDetailHeightSmooth(nodeId.ToNode().m_position);
+            if (!viewOnly) {
+                RenderLaneOverlays(cameraInfo);
+                RenderFloatingLaneCurve(cameraInfo);
+            }
+        }
 
-                LaneEndTransitionGroup groupAtNode = group_;
-                if (nodeId != SelectedNodeId) {
-                    if (AltIsPressed) {
-                        groupAtNode = LaneEndTransitionGroup.Track;
+        private void RenderNodeCircles(RenderManager.CameraInfo cameraInfo) {
+            if (GetSelectionMode() == SelectionMode.None) {
+                for (int cacheIndex = CachedVisibleNodeIds.Size - 1; cacheIndex >= 0; cacheIndex--) {
+                    var nodeId = CachedVisibleNodeIds.Values[cacheIndex];
+                    bool isNodeVisible = MainTool.IsNodeVisible(nodeId);
+                    if (nodeId == HoveredNodeId && isNodeVisible) {
+                        Highlight.DrawNodeCircle(
+                            cameraInfo: cameraInfo,
+                            nodeId: this.HoveredNodeId,
+                            color: this.MainTool.GetToolColor(warning: Input.GetMouseButton(0), error: false),
+                            alpha: true,
+                            overrideRenderLimits: true);
                     } else {
-                        groupAtNode = LaneEndTransitionGroup.Vehicle;
+                        Highlight.DrawNodeCircle(
+                        cameraInfo: cameraInfo,
+                        nodeId: nodeId,
+                        color: isNodeVisible ? DefaultLaneEndColor : DefaultDisabledLaneEndColor,
+                        alpha: true);
                     }
                 }
+            }
+        }
 
-                foreach (LaneEnd laneEnd in laneEnds) {
-                    ref NetLane sourceLane = ref laneEnd.LaneId.ToLane();
-                    if (!sourceLane.IsValidWithSegment()) {
-                        continue;
-                    }
+        private void RenderLaneCurves(RenderManager.CameraInfo cameraInfo) {
+            for (int cacheIndex = CachedVisibleNodeIds.Size - 1; cacheIndex >= 0; cacheIndex--) {
+                var nodeId = CachedVisibleNodeIds.Values[cacheIndex];
+                if (MainTool.IsNodeVisible(nodeId)) {
```

Comment:
If we cache visible nodes, do we need to check visibility here again?

Reply:
Not sure. I just copy pasted the code into a method. Didn't try to make changes were possible. less chance of introducing new errors.

Reply:
is deep nesting a bad thing?
If you don't like it I can change the code to how it was before: use return instead of nesting. 
Some people would argue having single exit point for the method is important. but that is usually for C++ to manage resources.

Reply:
Harder to read, much harder in fact, and we read code a lot

Reply:
return approach it is then.